### PR TITLE
Fix uncached Windows CI

### PR DESCRIPTION
**Pull request by jonas-chevalier-da**
UPDATE (@aherrmann-da): Migration to the GCP Windows pool (https://github.com/digital-asset/daml/pull/599) triggered issues with missing standard header files when building the boringssl target. These could be reproduced on the Azure hosted nodes by disabling the remote cache.

This PR changes the CC toolchain setup on Windows from the old CROSSTOOL mechanism to the new [cc_toolchain_config mechanism](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html). With these changes an uncached build [succeeds on the Azure hosted nodes](https://dev.azure.com/digitalasset/daml/_build/results?buildId=4741).

Additionally the haskell_zlib target is fixed by adding the `includes` attribute. See https://github.com/tweag/rules_haskell/pull/861.

---

Used to be "ci: test the windows build without caching"

While trying to use the new windows pool ( #599 ) we encountered this issue:
```

ERROR: C:/users/vssadministrator/_bazel_vssadministrator/w3d6ug6o/external/boringssl/BUILD:147:1: C++ compilation of rule '@boringssl//:ssl' failed (Exit 1): gcc failed: error executing command 
  cd C:/users/vssadministrator/_bazel_vssadministrator/w3d6ug6o/execroot/com_github_digital_asset_daml
  SET PATH=C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\usr\bin;C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\bin
    SET PWD=/proc/self/cwd
  external/io_tweag_rules_haskell_ghc_windows_amd64/mingw/bin/gcc -std=gnu++0x -MD -MF bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/tls13_server.d -frandom-seed=bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/tls13_server.o -iquote external/boringssl -iquote bazel-out/x64_windows-fastbuild/genfiles/external/boringssl -iquote bazel-out/x64_windows-fastbuild/bin/external/boringssl -isystem external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/genfiles/external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/bin/external/boringssl/src/include -DGRPC_BAZEL_BUILD -DWIN32_LEAN_AND_MEAN -DOPENSSL_NO_ASM -no-canonical-prefixes -fno-canonical-system-headers -c external/boringssl/src/ssl/tls13_server.cc -o bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/tls13_server.o
Execution platform: @bazel_tools//platforms:host_platform

In file included from external/boringssl/src/include/openssl/base.h:379:0,
                 from external/boringssl/src/include/openssl/ssl.h:145,
                 from external/boringssl/src/ssl/tls13_server.cc:22:
C:\\users\\vssadministrator\\_bazel_vssadministrator\\w3d6ug6o\\execroot\\com_github_digital_asset_daml\\external\\io_tweag_rules_haskell_ghc_windows_amd64\\mingw\\bin/../lib/gcc/x86_64-w64-mingw32/7.2.0/../../../../include/c++/7.2.0/memory:67:10: fatal error: bits/stl_raw_storage_iter.h: No such file or directory
 #include <bits/stl_raw_storage_iter.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Here we are testing if the issue is reproducible with the MS-hosted agent, assuming that it's an issue with the cache.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by flokli** [comment]
rebased onto latest `master` (with the GHC upgrade as suggested by @moritzkiefer-da)

cc @aherrmann-da

**Comment by flokli** [comment]
With the latest GHC, things still fail:

```
SUBCOMMAND: # @boringssl//:ssl [action 'Compiling external/boringssl/src/ssl/ssl_cert.cc']
cd C:/users/vssadministrator/_bazel_vssadministrator/w3d6ug6o/execroot/com_github_digital_asset_daml
  SET PATH=C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\usr\bin;C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\bin
    SET PWD=/proc/self/cwd
  external/io_tweag_rules_haskell_ghc_windows_amd64/mingw/bin/gcc -std=gnu++0x -MD -MF bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_cert.d -frandom-seed=bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_cert.o -iquote external/boringssl -iquote bazel-out/x64_windows-fastbuild/genfiles/external/boringssl -iquote bazel-out/x64_windows-fastbuild/bin/external/boringssl -isystem external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/genfiles/external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/bin/external/boringssl/src/include -DGRPC_BAZEL_BUILD -DWIN32_LEAN_AND_MEAN -DOPENSSL_NO_ASM -no-canonical-prefixes -fno-canonical-system-headers -c external/boringssl/src/ssl/ssl_cert.cc -o bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_cert.o

ERROR: C:/users/vssadministrator/_bazel_vssadministrator/w3d6ug6o/external/boringssl/BUILD:147:1: C++ compilation of rule '@boringssl//:ssl' failed (Exit 1): gcc failed: error executing command 
  cd C:/users/vssadministrator/_bazel_vssadministrator/w3d6ug6o/execroot/com_github_digital_asset_daml
  SET PATH=C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\usr\bin;C:\Users\VssAdministrator\dadew\scoop\apps\msys2-20180531\current\bin
    SET PWD=/proc/self/cwd
  external/io_tweag_rules_haskell_ghc_windows_amd64/mingw/bin/gcc -std=gnu++0x -MD -MF bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_aead_ctx.d -frandom-seed=bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_aead_ctx.o -iquote external/boringssl -iquote bazel-out/x64_windows-fastbuild/genfiles/external/boringssl -iquote bazel-out/x64_windows-fastbuild/bin/external/boringssl -isystem external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/genfiles/external/boringssl/src/include -isystem bazel-out/x64_windows-fastbuild/bin/external/boringssl/src/include -DGRPC_BAZEL_BUILD -DWIN32_LEAN_AND_MEAN -DOPENSSL_NO_ASM -no-canonical-prefixes -fno-canonical-system-headers -c external/boringssl/src/ssl/ssl_aead_ctx.cc -o bazel-out/x64_windows-fastbuild/bin/external/boringssl/_objs/ssl/ssl_aead_ctx.o
Execution platform: @bazel_tools//platforms:host_platform

In file included from external/boringssl/src/include/openssl/base.h:379:0,
                 from external/boringssl/src/include/openssl/ssl.h:145,
                 from external/boringssl/src/ssl/ssl_aead_ctx.cc:15:
C:\\users\\vssadministrator\\_bazel_vssadministrator\\w3d6ug6o\\execroot\\com_github_digital_asset_daml\\external\\io_tweag_rules_haskell_ghc_windows_amd64\\mingw\\bin/../lib/gcc/x86_64-w64-mingw32/7.2.0/../../../../include/c++/7.2.0/memory:67:10: fatal error: bits/stl_raw_storage_iter.h: No such file or directory
 #include <bits/stl_raw_storage_iter.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

File: bazel_tools/haskell-win-sys-includes.patch [hunk 1]
```diff
@@ -13,15 +13,19 @@ Generated with:
...
+index ae16776..022e15b 100644
 --- a/haskell/CROSSTOOL.windows
 +++ b/haskell/CROSSTOOL.windows
-@@ -30,6 +30,8 @@ toolchain {
+@@ -30,7 +30,12 @@ toolchain {
    tool_path { name: "objcopy" path: "mingw/bin/objcopy" }
    tool_path { name: "objdump" path: "mingw/bin/objdump" }
    tool_path { name: "strip" path: "mingw/bin/strip" }
 +
 +  cxx_builtin_include_directory: "C:/"
```
Comment (flokli) [inline bazel_tools/haskell-win-sys-includes.patch]: why C:/ ?

Comment (majcherm-da) [reply bazel_tools/haskell-win-sys-includes.patch]: https://github.com/digital-asset/daml/issues/180

**Comment by flokli** [comment]
I did revert the windows-specific `fno-canonical-` flags in `rules_haskell`, and boringssl builds again.

**Comment by aherrmann-da** [comment]
Pushed changes in Hazel that perform `./configure` in a build rule rather than repository rule, to see if this allows to remove `-fno-canonical-...` to work around the boringssl issue.

**Comment by aherrmann-da** [comment]
- Using `cc_toolchain_config` instead of `CROSSTOOL` following [this tutorial](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html).
- Dropping `-no-canonical-*` flags
- Fixing `haskell_zlib` by replacing `strip_include_prefix` with `includes`.
With these changes `boringssl` and `da-hs-base` build.

**Comment by neil-da** [comment]
@aherrmann-da / @garyverhaegen-da  - are one of you going to take over this PR?

**Comment by aherrmann-da** [comment]
@neil-da Yes, happy to take it over. To clarify, the purpose of this PR was to reproduce and resolve the issues that were encountered in #599. It looks like we've found a solution. I'll clean it up and get it ready for merge.

File: bazel_tools/haskell-cc-toolchain.patch [hunk 2]
```diff
@@ -0,0 +1,238 @@
+diff --git a/haskell/CROSSTOOL.windows b/haskell/CROSSTOOL.windows
```
Comment (cocreature) [inline bazel_tools/haskell-cc-toolchain.patch]: Does it make sense to upstream this patch?

File: bazel_tools/hazel-cc-toolchain.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff --git a/hazel/third_party/haskell/BUILD.zlib b/hazel/third_party/haskell/BUILD.zlib
```
Comment (cocreature) [inline bazel_tools/hazel-cc-toolchain.patch]: Same question: Should this be upstreamed?

File: bazel_tools/haskell-cc-toolchain.patch (hunk 2 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-cc-toolchain.patch]: Yes, I think so. I'll upstream it.

**Comment by aherrmann-da** [comment]
Closed in favor of #1047 as mergify rejected this because opening user needs to be member.

**Pull request closed without merging**